### ISSUE TITLE
Error Bubbling

### ISSUE THREAD

**Issue by lippling** [slamdata/purescript-halogen#278]
Hi,

I started using Purescript/Halogen a couple of days ago, but I have used Haskell before.

Component structure:

```
         Main
      /       \
Login           App
               /    \
      Navigation     Content
                      / | \
                      .....     
```

The app talks to an API with an access token. When access token and refresh token are expired, the user should see the login again. So for example: a child component of the `Content` component makes an Ajax call and the response is `401`. I'd like to have one point in `Main` where I can "catch" all `401` errors, so I can install the `Login` component instead of `App`.

Is this possible or do I have to deep `peek` every single child component that does Ajax or is there a completely different approach?

Thanks
Alexander

**Comment by natefaubion** [comment]
There's no way currently to notify a parent from a child except through passively peeking. Something like #245 might help a little.

One possibility thought is to assemble an HTTP service at your root that lets you intercept these kinds of errors, and then use the `subscribe'` mechanism to get notified when an error happens. You then propagate your service down to your other components.

There are lots of ways to do the propagation, but they all basically amount to some sort of argument passing. The simplest is to just use a record for you environment, and turn your `Component`s into functions of `Env -> Component`. Or take advantage of structural typing and something like `forall s. { http :: MyHTTPService | s } -> Component`.

If you took it a step further, you could use `Reader` as a sort of dependency injection monad for configuring your components.

```purescript
type Env =
  { http :: HTTPService
  , notify :: NotificationService
  , rpc :: RPCService
  }

mainComponent :: Reader Env (Component ...)
mainComponent = do
  sidebar <- sidebarComponent
  items <- itemComponent
  pure $ component ...

sidebarComponent :: forall s. Reader { http :: HTTPService | s } (Component ...)

itemComponent :: forall s. Reader { notify :: NotificationService | s } (Component ...)
```

**Comment by lippling** [comment]
What do you mean with HTTPService and  `subscribe'`? As far as I understood, this is intended to integrate JS code.
I just want to use `purescript-affjax` for my API calls.

**Comment by natefaubion** [comment]
You can tell me if I'm wrong, but it sounds like you are wanting to globally configure the behavior of your HTTP calls to whatever backend you are using, so that anytime you get an auth error (regardless of which component is issuing the request), your app will throw up a login box.

This is something that can't be solved by peeking because peeking only lets you observe an input, it doesn't let you observe the intermediate result of that input. So while you can peek that some component got the input `GetItems`, you can't observe that it resulted in an HTTP error.

Halogen also does not provide a way for you to globally address components, because this breaks modularity. You can't say "on error, send an input to the root component" from within a child component, because that would require the child to know where it is installed.

One way to create this loop is at the root when you `runUI` your app. You get a driver that lets you push in new inputs from outside of your app. You can configure your own `affjax` function that, when faced with an error, pushes a new input into the driver. You then pass that custom `affjax` function through your hierarchy for components to use (instead of calling the default `Network.HTTP.Affjax.affjax`).

```purescript
main = onLoad $ runAff throwException (const (pure unit)) do
  driverVar <- makeVar
  let myAffjax = mkAffjaxDriver driverVar
  { node, driver } <- runUI (ui myAffjax) (installedState initialState)
  putVar driverVar driver

  where
  mkAffjaxDriver
    :: forall a b
     . (Requestable a, Respondable b)
    => AVar (Driver MyQuery _)
    -> AffjaxRequest a
    -> Affjax _ (AffjaxResponse b)
  mkAffjaxDriver driverVar req = do
    resp <- req
    case resp.statusCode of
      StatusCode 401 -> do
        driver <- takeVar driverVar
        putVar driverVar driver 
        driver (left $ action AuthFailed)
      _ ->
        pure unit
    pure resp
```

I have not compiled this code, but I think something like this would work (barring type and eff errors). I mentioned `subscribe'` in case you wanted to be notified in multiple places. You could set up some sort of event emitter type thing when creating your app and pass that around.

The point is, if you want to configure app-specific behavior like this, you are going to have to pass things around through your app. The rest of the previous post was just a way you might pass things around that you've configured. If you have multiple "services" that you've configured (maybe it's a custom `affjax` handler like above, or maybe its a record of functions configured for a specific user), you'll probably want to put it all in one record, and that way you can use structural typing to only list what you care about.

**Comment by zudov** [comment]
I think the best way to handle it would be to make an algebra for http queries and then give a free monad (transformer) derived from it as a `g` argument to your `Component`. Then, before you do `runUi` you would have to `interpret` your component transforming its `g` parameter to an `Aff`. The interpreter would intercept  responses with `status == 401` and issue a query using the driver.

I don't have an example of doing it with http queries, and I don't think anyone made a free monad interface to affjax yet, however I've just been working on doing the same for WebStorage API, [here is an example](https://github.com/zudov/purescript-webstorage-free/blob/freet-identity/examples/halogen-storage/src/Main.purs) which uses WebStorage in the same way as I described above.

In `eval` I "initiate" a query in a `Storage` monad, [here](https://github.com/zudov/purescript-webstorage-free/blob/freet-identity/examples/halogen-storage/src/Main.purs#L57), but the actual execution happens in my `main` [here](https://github.com/zudov/purescript-webstorage-free/blob/freet-identity/examples/halogen-storage/src/Main.purs#L62). I do `(interpret (runStorage localStorage) ui)`, I could pass `sessionStorage` there and then all the interaction would be done with `sessionStorage` without me having to update the evals. In the same way I could pass any interpreter there and such interpreter could intercept error responses and report them to the component which would show them to user.

**Comment by zudov** [comment]
There is also an [`interpret` example](https://github.com/slamdata/purescript-halogen/blob/master/examples/interpret/src/Main.purs#L54) which demonstrates that technique. You can change the [`evalOutput`](https://github.com/slamdata/purescript-halogen/blob/master/examples/interpret/src/Main.purs#L54) to do whatever you want whenever `output` is executed. I think that is a much cleaner way of injecting the customised behaviour.

**Comment by natefaubion** [comment]
@zudov I think this gets to the different ways of handling dependencies: you can either take it as an argument, or you can interpret a result. If all you want to do is make some Affjax requests, I think passing around modules is a great, fairly simple way to do it.

To wit, free monads have their own set of limitiations. While you gain flexibility in writing an interpretter, it can also require a lot more machinery. For example, if I used a free monad to for HTTP, I'd be in a bind if I then needed to make parallel requests (I think you'd also need to provide a separate free applicative for these purposes, because you can't have both a free applicative and monad).

I personally don't mind keeping most components in `Aff` and just passing around modules. The great thing about `interpret` is down the line you can always refactor it with a free monad and interpret it locally if you need to.

In any case, you still need to do all the nasty work of tying the knot with the driver (the interpreter needs the driver, but you can't get the driver until you've applied the interpreter). Maybe we can provide a better way to do that?

**Comment by zudov** [comment]
@natefaubion That's a great explanation, thanks. I should give a try to this 'driver' pattern.

**Comment by natefaubion** [comment]
@zudov I wouldn't call module passing a 'driver' pattern. You'd have to do the same gymnastics with the driver using a custom interpreter. `interpret` only lets you swap out the underlying functor for a component, while the crux of this use case is that you have to be able to feed the driver while interpreting. Halogen has 3 ways to feed the driver:

1. DOM events
2. `subscribe`
3. Using the `driver` obtained from `runUI`

Since 1 is obviously out, you either have to do it with `subscribe` or at the root of your application (after you've already run it). It would be nice if there was an easier way to tie the knot (maybe https://github.com/zrho/purescript-monad-fix would work?).

**Comment by lippling** [comment]
@natefaubion @zudov Many thanks for your input. It helped a lot!

I didn't look into the `Free` approach, because a) I only have a vague notion of the `Free` monad, b) I wouldn't be able to implement such a thing and c) there seem to be some caveats, as @natefaubion mentioned.

So I took @natefaubion's approach and came up with this:

```
type QueryP = Coproduct Query (ChildF ChildSlot ChildQuery)
type AppEffects eff = (ajax :: AJAX, console :: CONSOLE | eff)
type Effects eff = HalogenEffects (AppEffects eff)

main :: Eff (Effects ()) Unit
main = runAff throwException (const (pure unit)) do
    driverVar <- makeVar
    let myAffjax = mkAffjaxDriver driverVar
    app <- runUI (ui myAffjax) (installedState initialState)
    putVar driverVar app.driver
    onLoad $ appendToBody app.node
    where
        mkAffjaxDriver :: AVar (Driver QueryP (AppEffects ()))
                        -> (forall a b. (Requestable a, Respondable b) => AffjaxRequest a
                        -> Aff (Effects ()) (AffjaxResponse b))
        mkAffjaxDriver driverVar req = do
            resp <- affjax req
            case resp.status of
                StatusCode 401 -> do
                    driver <- takeVar driverVar
                    driver (left $ action AuthFailed)
                    putVar driverVar driver
                _ -> pure unit
            pure resp
```
I changed the type signature of `mkAffjaxDriver`, because `forall a b.` was in the wrong place.

**Comment by jdegoes** [comment]
> For example, if I used a free monad to for HTTP, I'd be in a bind if I then needed to make parallel requests (I think you'd also need to provide a separate free applicative for these purposes, because you can't have both a free applicative and monad).

You can, however, create a separate algebra to support parallelizing operations, which is then expressed inside free, for an overall structure of a sequential flow of parallel tasks that rejoin on every bind.

On the topic of this issue, right now, the component machinery allows you to observe child inputs, but not child outputs. Moreover, it requires parent and child share the same `g`.

However, neither of these are necessary limitations. If we generalize this machinery (at the very leastm the first part; the second would be helpful here but is not necessary), you would be able to "peek" not just on inputs, but also on outputs. 

I feel like that could head off a good amount of complexity / anti-patterns.

Thoughts? /cc @garyb

**Comment by zudov** [comment]
If component could query itself (in such a way that parent can observe that query with `peek`, #257) that would be one way of peeking on child's output. Did I get that right?

For example, whenever a component gets an http error it would do something like `queryItself (action GotAnError)`. And whenever peeking parent observes `GotAnError` it would do what needs to be done.

**Comment by garyb** [comment]
I've not caught up with this thread properly yet but will do soon and try to offer something of use!

@zudov a component can sort-of query itself already, through `subscribe`. I've lately been musing a way of using the subscribe/coroutine method as an alternative to peeking (and raising of HTML events), so they're unified under the one mechanism, as I think there might be a relatively easy way of doing it.

**Comment by jdegoes** [comment]
@zudov That's not necessary, it's actually orthogonal (although having a way to do "loopback" would also solve various types of problems like this). 

Rather, it would be a way to intercept a child's `g a` output (simplified type sig), and transform that to a value of the same type in the parent's `g` (note that right now, parent and child have same `g`, but I generalized here just for the sake of argument). The default for this would be the identity transformation, but you could emit a predecessor action or a successor action in the parent's `g`, so long as you satisfied the natural transformation requirement to emit the `a` promised by the child.

This would allow you to "catch" errors and such, and if the components are all outputing to something like `Aff`, then anywhere along the stream you can catch `Aff` errors and deal with them (or rethrow them, too).

**Comment by natefaubion** [comment]
With the current architecture it's easy to insert your own base effect type which can be used to handle things like global authentication errors. We also have first-class support for output messages from components (though for this use case, a custom effect type would be best). With subscriptions and something like `purescript-aff-bus`, one can handle global events for things like error notifications.

**Comment by denibertovic** [comment]
@natefaubion Is there an example somewhere where custom effect types or purescript-aff-bus are used to handle affjax exceptions from components? What is the general best practice for doing this? Or is the `driver` approach mentioned above still the way to go?
